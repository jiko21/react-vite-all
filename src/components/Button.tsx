/** @jsxImportSource @emotion/react */
import { css } from '@emotion/react';
import React, { ButtonHTMLAttributes } from 'react';

export type ButtonStyle = 'primary' | 'danger';

type ButtonProps = ButtonHTMLAttributes<HTMLButtonElement> & {
  buttonStyle?: ButtonStyle;
  disabled?: boolean;
  label: React.ReactNode;
};

const baseStyle = css`
  border: none;
  border-radius: 4px;
  font-weight: 700;
  height: 32px;
  line-height: 16px;
  min-width: 60px;
  padding: 8px 4px;
  position: relative;
  :active {
    top: 1px;
  }
`;

const primaryStyle = css`
  background-color: #00c1cf;
  color: white;
  :hover {
    background-color: #5ec9d1;
    color: white;
  }
`;

const dangerStyle = css`
  background-color: #ff002b;
  color: white;
  :hover {
    background-color: #ff3b3b;
    color: white;
  }
`;

const disabledStyle = css`
  background-color: #dfdfdf;
  color: #3c3c3c;
  :active {
    top: 0;
  }
`;

export const Button: React.FC<ButtonProps> = ({
  buttonStyle = 'primary',
  disabled = false,
  label,
  ...props
}) => {
  const style = disabled ? disabledStyle : (buttonStyle === 'danger' ? dangerStyle : primaryStyle);
  return <button css={[baseStyle, style]}>{label}</button>;
};
