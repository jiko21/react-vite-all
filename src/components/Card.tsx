/** @jsxImportSource @emotion/react */
import { css } from '@emotion/react';
import React from 'react';

const style = css`
  background-color: #eeeeee;
  border-radius: 4px;
  padding: 8px;
`;

export const Card: React.FC = ({ children }) => <div css={style}>{children}</div>;
