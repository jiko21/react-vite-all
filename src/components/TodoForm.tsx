import React, { useMemo, useState } from 'react';
import { Button } from './Button';
import { Input } from './Input';

type Props = {
  handleSubmit: (text: string) => void;
};

export const TodoForm: React.FC<Props> = ({ handleSubmit }) => {
  const [text, setText] = useState('');
  const disabled = useMemo(() => text.length === 0, [text]);
  const onSubmit = () => {
    if (text.length > 0) {
      handleSubmit(text);
      setText('');
    }
  };

  return (
    <div>
      <Input type="text" value={text} onChange={(e) => setText(e.target.value)} testId="todoForm" />
      <Button onClick={onSubmit} disabled={disabled} label="ADD" testId="todoForm" />
    </div>
  );
};
