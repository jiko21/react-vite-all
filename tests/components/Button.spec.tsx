import { describe, expect, it, vi } from 'vitest';
import { fireEvent, render, screen } from '@testing-library/react';
import React from 'react';
import { Button } from '../../src/components/Button';

describe('components/Button', () => {
  it('correctly render: primary', () => {
    const component = render(<Button label={'sample'} />);
    expect(component.container).toMatchSnapshot();
  });

  it('correctly fire event: primary', () => {
    const testId = 'unit';
    const onClickMock = vi.fn();
    render(<Button label={'sample'} testId={testId} onClick={onClickMock} />);
    fireEvent.click(screen.getByTestId(`${testId}-button`));
    expect(onClickMock).toBeCalled();
  });

  it('correctly render: danger', () => {
    const component = render(<Button label={'sample'} buttonStyle="danger" />);
    expect(component.container).toMatchSnapshot();
  });

  it('correctly fire event: danger', () => {
    const testId = 'unit';
    const onClickMock = vi.fn();
    render(<Button label={'sample'} testId={testId} buttonStyle="danger" onClick={onClickMock} />);
    fireEvent.click(screen.getByTestId(`${testId}-button`));
    expect(onClickMock).toBeCalled();
  });

  it('correctly render: disabled', () => {
    const component = render(<Button label={'sample'} disabled />);
    expect(component.container).toMatchSnapshot();
  });

  it('correctly not fire event: disabled', () => {
    const testId = 'unit';
    const onClickMock = vi.fn();
    render(<Button label={'sample'} testId={testId} disabled onClick={onClickMock} />);
    fireEvent.click(screen.getByTestId(`${testId}-button`));
    expect(onClickMock).not.toBeCalled();
  });
});
