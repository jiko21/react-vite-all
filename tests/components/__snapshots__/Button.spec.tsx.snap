// Vitest Snapshot v1

exports[`components/Button > correctly render: danger 1`] = `
<div>
  <button
    class="css-gu30kj-Button"
    data-testid="undefined-button"
  >
    sample
  </button>
</div>
`;

exports[`components/Button > correctly render: disabled 1`] = `
<div>
  <button
    class="css-5shtp7-Button"
    data-testid="undefined-button"
  >
    sample
  </button>
</div>
`;

exports[`components/Button > correctly render: primary 1`] = `
<div>
  <button
    class="css-t9php5-Button"
    data-testid="undefined-button"
  >
    sample
  </button>
</div>
`;
